### ISSUE TITLE
Ruby 2.4 Compatible?

### ISSUE THREAD

**Issue by expandrive** [eventmachine/eventmachine#806]
The Readme claims support up to Ruby 2.3 - I wanted to sort out if there were any known issues with 2.4 or if the Readme needs to get updated.

**Comment by kenpropel** [comment]
@expandrive not that I've done extensive testing, but we're running on Ruby 2.4 without any issues. We're using `thin` and `em-http-request` that both depend on `eventmachine`.

**Comment by sodabrew** [comment]
Works fine with Ruby 2.4. I'll make a note to update the README.

Note that Windows binaries are not available for Ruby 2.4 yet, in that case you'll need to compile yourself, which is easy with: `gem install eventmachine --platform ruby`

**Comment by hemancuso** [comment]
Great, thanks!

**Comment by MSP-Greg** [comment]
@hemancuso 

See PR #808.  Until it's accepted, you need to make sure your MSYS2/MinGW system has the OpenSSL package installed, otherwise EM will compile without SSL support.

**Comment by sodabrew** [comment]
Thanks @MSP-Greg! Merged!

**Comment by MSP-Greg** [comment]
@sodabrew

I forgot about something.  As you know, ruby compiles OpenSSL and Digest with an OpenSSL package.  OpenSSL is available as a gem, Digest will be a gem in the near (?) future (because of this issue).

The potential exists to compile EM with a different version of the OpenSSL package than what was used for the Ruby build.  At present, I'm not sure if anyone is really sure whether this is a problem or not.

MinGW is currently at 1.0.2l, and Ruby 2.4.2 was built with the same version.  I know some versions of 2.4.1 were built with 1.0.2k.

Have you seen any problems with this?  I would hope that 1.0.2 versions are API compatible, but I don't know.  Once MinGW moves to 1.1.0, things might get a bit messier...

**Comment by steve-misky** [comment]
FWIW, I've just upgraded from Ruby 2.3.1 to Ruby 2.4.2 with openssl 1.1.0f - both centos7 and Mac with no issues.  Of course it took a few months for RVM and things to stabilize, but as of Mid September things got ironed out.

We use EventMachine w/em-http-request (along with http2 GEM for HTTP/2)

I installed openssl from source (both systems) and for Linux had to fix a unit-test to get install to work.
On MAC, we needed to brew install openssl@1.1 (even though we linked with openssl build libs).
The "trick" appeared to be providing the correct PATH on the rvm install to where the openssl Libs are.

We are able to use TLSv1.2 w/PolyChaCha ciphers (which was the goal along with prepping for TLS1.3 release from openssl


#####  Mac OS #####
NOTE: we install the new SSL Libs in /usr/local/ssl, and point RVM to that same location
            using 2 different flags (which seemed to be the key to get it to work)

brew install openssl@1.1

# if needed
(sudo) mkdir /usr/local/src
(sudo) chmod 777

(As yourself)
cd /usr/local/src
curl --remote-name https://www.openssl.org/source/openssl-1.1.0f.tar.gz
tar -xvf openssl-1.1.0f.tar.gz
cd openssl-1.1.0f
./config --prefix=/usr/local/ssl enable-weak-ssl-ciphers 
enable-deprecated enable-rc4 enable-ssl3 enable-ssl3-method

make
make test
(sudo) make install

$ rvm get stable
  
$ rvm install 2.4.2 --with-openssl-dir=/usr/local/ssl --with-ssl-dir=/usr/local/ssl
$ gem install bundler
$ bundler install # (because we have a Gemfile with all the GEMs/Versions we require)


######   CENTOS7   ######
# New OPENSSL unit-tests require perl core packages (the error you'll get is shown here)
# TO AVOID these errors when running make test
#         "Can't locate Module/Load/Conditional.pm in (....)"
$ (sudo) yum install perl-core

(as root)
cd /usr/local/src
wget https://openssl.org/source/openssl-1.1.0f.tar.gz
tar \-xzf openssl-1.1.0f.tar.gz
cd openssl-1.1.0f
./config shared enable-weak-ssl-ciphers enable-deprecated enable-rc4 enable-ssl3 enable-ssl3-method
make
make test
make install

# will put the compiled version in  /usr/local/ssl (lib/include/bin etc.)
# BUT does NOT put them in the /usr/lib64 directory (which is where openssl is looking)

$ /usr/local/ssl/bin/openssl version
    openssl: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory

# I then had to link /usr/lib64/libssl.so.1.1 to /usr/local/ssl/lib/libssl.so.1.1
# and also the libcrypto.so.1.1

# (NOW VERIFY the new openssl can LINK with the new libs) -  SUCCESS!
$ /usr/local/ssl/bin/openssl version
OpenSSL 1.1.0f  25 May 2017

# As yourself (in a shared setup)
$ rvm install 2.4.2 --with-openssl-dir=/usr/local --with-ssl-dir=/usr/local
$ gem install bundler
$ bundle install   # Again, I have a Gemfile

**Comment by neongrau** [comment]
Could someone explain what's wrong here. I got a Windows server where i just upgraded to ruby 2.4.2 and got the development kit installed.
Since i had problems with eventmachine i ran `gem install eventmachine --platform ruby`. Native extension got built and everything seemed fine.

I now wanted to deploy that ruby to a server that hasn't got the devkit installed and eventmachine complains again `The specified module could not be found.   - C:/ruby/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.5/lib/rubyeventmachine.so (LoadError)`.

**Comment by MSP-Greg** [comment]
@neongrau

I suspect it's due to a missing `libwinpthread-1.dll` file, which should be placed in `bin\ruby_builtin_dlls`.  I don't know if it will need to be added to `ruby_builtin_dlls.manifest`.

I do almost all of my work with a current trunk build, and my build handles dll's differently than standard RubyInstaller2.

Regardless, when adding extension gems to a ruby install without MSYS2/MinGW, checking the *.so files for dll dependencies with something like PEStudio is often helpful...

**Comment by neongrau** [comment]
@MSP-Greg 

Thanks a lot! That did it, had to place the dll into `bin` directly though.

**Comment by MSP-Greg** [comment]
@neongrau Glad that did it.

I'd be interested to know if it works if you were to add it to `bin\ruby_builtin_dlls`, then change `ruby_builtin_dlls.manifest` to be:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
  <assemblyIdentity type="win32" name="ruby_builtin_dlls" version="1.0.0.0"></assemblyIdentity>
  <file name="libeay32.dll"/>
  <file name="libffi-6.dll"/>
  <file name="libgdbm-4.dll"/>
  <file name="libgdbm_compat-4.dll"/>
  <file name="libgmp-10.dll"/>
  <file name="libiconv-2.dll"/>
  <file name="libintl-8.dll"/>
  <file name="libwinpthread-1.dll"/>
  <file name="libyaml-0-2.dll"/>
  <file name="ssleay32.dll"/>
  <file name="zlib1.dll"/>
</assembly>
```

On my system, with the above, if I rename `msys64\mingw64\bin` to `bin_save`, then add an empty `bin` folder (thereby removing any dll's from the MSYS2/MinGW path), the following works:

```
ruby -reventmachine -e "puts EM.ssl?"
```

I hate adding dll's to bin...

@sodabrew sorry for getting a bit off-topic,..

**Comment by sodabrew** [comment]
Where does `libwinpthread-1.dll` come from if it's not already installed in the Ruby libs path?

**Comment by MSP-Greg** [comment]
On a system that is capable of compiling EM, it's contained in the MSYS2/MinGW installation, which also contains base headers and library files, along with the same for any installed packages.

The dll locations in the MSYS2/MinGW installation are added to PATH by the builds thru the rubygems require hooks.  But, if the tools aren't installed...

Where it can get messy (on Windows) is duplicate files.  For instance, many of the MSYS2/MinGW dll's also exist in 'Git for Windows' installs.  Worse yet, many duplicate app files (*.exe) exist in 'MSYS2/MinGW', 'Git for Windows', and the new Windows 'bash system'.  There are probably more, as a lot of '*nix' software uses the same build tools.

This is why some people are rather 'path conscious' in appveyor.yml files.  Same reason that RubyInstaller2 builds use windows specific os calls to add to the dll path of ruby.

Hope this makes sense...

**Comment by MSP-Greg** [comment]
@sodabrew

1. I know nothing about OSX or *nix OS's.  Conversely, I've been using Windows and coding on Windows for a *long* time (I used & coded on DOS).  So, I may leave out something that I shouldn't, or I may over-explain something under the (possibly mistaken) assumption that some OSX or *nix users know as much about windows as I know about their OS's.

2. Ruby builds 2.3 and below were compiled using a custom build system that also required the maintainers to build custom packages.  Packages were available for most popular extension gems.

3. Ruby builds for 2.4 and later are using MSYS2/MinGW tools / packages.  It's a system that is used by a lot of c/c++ apps, along with several other languages.  The repo for it has over 1,000 packages.

4. My MSYS2/MinGW build system is about 2.5 gig in size.  It isn't a standard part of windows.  The runtime files included with windows ruby will allow a 'new user' to download/install it, but it's not part of install package.  Some gems may require additional packages to compile, that's what the gemspec metadata hook does, so user's don't need to know how to do so.

5. Critical item I left out.  Windows Ruby builds are self-contained, in that they contain all the dll files needed to run core and std-lib.

`libwinpthread-1.dll` is a grey area dll, in that ruby doesn't seem to require it, but it may be required by code compiled with only the standard tool-chain and packages used by ruby.  I include it in my trunk builds, but the standard Windows ruby builds do not.  The issue of 'compile on system X (with build tools)' and 'move/copy to system Y (without build tools)' hasn't really been addressed anywhere in a Wiki...

Sorry about the long explanation., and thanks.

**Comment by neongrau** [comment]
@MSP-Greg tried with the dll in `ruby_builtin_dlls` and adding it to `ruby_builtin_dlls.manifest` but that didn't work. Same as if the dll was missing.

`bin` previously had one single dll namely `x64-msvcrt-ruby240.dll` which is also used by eventmachine.so as PEStudio had shown. So that's why i just put it into there too.

Maybe when building eventmachine the dll should be linked statically since it's not used anywhere else?

**Comment by MSP-Greg** [comment]
@neongrau

Thanks for checking.  I need to check that again, as it seemed to work on my system.  After all, EM is finding the openssl dll's in the `ruby_builtin_dlls` folder...

> Maybe when building eventmachine the lib should be linked statically since it's not used anywhere else?

There may be other gems that use it.   Not the best example, but the openssl package is used by Ruby digest and openssl, EM, and Puma.  There's probably more...

**Comment by todthomson** [comment]
Key is (on Windows + Ruby 2.4) to make sure you remove all versions of `eventmachine` e.g.

```

**Comment by sodabrew** [comment]
Resolved by #833 and 569c55ea3d58b8675e5fab045bccaea84d6f1215

**Comment by SteveALee** [comment]
Using latest install on Windows cmd today and get this error with ```jekyll --livereload```, OK without.

```gem install eventmachine --platform ruby``` doesn make any difference
